plugins/in_podman_metrics/podman_metrics.c

```diff
@@ -0,0 +1,496 @@
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2015-2022 The Fluent Bit Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_input_plugin.h>
+#include <fluent-bit/flb_config.h>
+#include <fluent-bit/flb_metrics.h>
+#include <fluent-bit/flb_metrics_exporter.h>
+#include <fluent-bit/flb_jsmn.h>
+
+#include <monkey/mk_core/mk_list.h>
+
+#include "podman_metrics.h"
+#include "podman_metrics_config.h"
+#include "podman_metrics_data.h"
+
+/*
+ * Collect information about podman containers (ID and Name) from podman configuration
+ * file (default is /var/lib/containers/storage/overlay-containers/containers.json).
+ * Since flb_jsmn library show JSON as a tree, search for objects with parent 0 (objects
+ * that are children to root array, and in them, search for ID and name (which is also
+ * an array.
+ */
+static int collect_container_data(struct flb_in_metrics *ctx)
+{
+    /* Buffers for reading data from JSON */
+    char buffer[CONFIG_READ_BUFF_SIZE];
+    char name[CONTAINER_NAME_SIZE];
+    char id[CONTAINER_ID_SIZE];
+
+    FILE *fp;
+
+    int array_id;
+    int r, i, j;
+    int read_bytes = 0;
+    int collected_containers = 0;
+
+    jsmn_parser p;
+    jsmntok_t t[JSON_TOKENS];
+
+    fp = fopen(ctx->config, "rb");
+    if (!fp) {
+        flb_plg_warn(ctx->ins, "Failed to open %s", ctx->config);
+        return -1;
+    }
+
+    read_bytes = fread(&buffer, sizeof(char), CONFIG_READ_BUFF_SIZE, fp);
+    buffer[read_bytes] = 0;
+    fclose(fp);
+    flb_plg_debug(ctx->ins, "Read %d bytes", read_bytes);
+
+    jsmn_init(&p);
+    r = jsmn_parse(&p, buffer, strlen(buffer), t, sizeof(t) / sizeof(t[0]));
+    if (r < 0) {
+        flb_plg_warn(ctx->ins, "Failed to parse JSON %d: %s", r, buffer);
+        return -1;
+    }
+
+    flb_plg_debug(ctx->ins, "Got %d nested tokens", t[0].size);
+
+    if (r < 1 || t[0].type != JSMN_ARRAY) {
+        flb_plg_warn(ctx->ins, "Expected array at the json root");
+        return -1;
+    }
+
+    for (i=0; i<r; i++) {
+        if (t[i].type == JSMN_STRING) {
+            if ((int)strlen(JSON_FIELD_ID) == t[i].end - t[i].start &&
+                strncmp(buffer + t[i].start, JSON_FIELD_ID, t[i].end - t[i].start) == 0) {
+                memset(id, 0, sizeof(id));
+                strncpy(id, buffer + t[i+1].start, t[i + 1].end - t[i + 1].start);
+                id[strlen(id)] = '\0';
```

Comment:
if `id` would not have a `\0` in place before, strlen would make it crash. I suggest to calculate the length of what you are copying in strncpy in a variable, e.g:

```c
len = t[i + 1].end - t[i + 1].start;
```

then just do `id[len] = '\0';`

with this approach, the memset() used before is not longer required.